sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeOneRowRelationSubquerySuite.scala

```diff
@@ -0,0 +1,161 @@
+import org.apache.spark.sql.catalyst.dsl.plans._
+import org.apache.spark.sql.catalyst.expressions.{Alias, ScalarSubquery}
+import org.apache.spark.sql.catalyst.plans._
+import org.apache.spark.sql.catalyst.plans.logical.{DomainJoin, LocalRelation, LogicalPlan, OneRowRelation, Project}
+import org.apache.spark.sql.catalyst.rules.RuleExecutor
+import org.apache.spark.sql.internal.SQLConf
+
+class OptimizeOneRowRelationSubquerySuite extends PlanTest {
+
+  private var optimizeOneRowRelationSubqueryEnabled: Boolean = _
+
+  protected override def beforeAll(): Unit = {
+    super.beforeAll()
+    optimizeOneRowRelationSubqueryEnabled =
+      SQLConf.get.getConf(SQLConf.OPTIMIZE_ONE_ROW_RELATION_SUBQUERY)
+    SQLConf.get.setConf(SQLConf.OPTIMIZE_ONE_ROW_RELATION_SUBQUERY, true)
+  }
+
+  protected override def afterAll(): Unit = {
+    SQLConf.get.setConf(SQLConf.OPTIMIZE_ONE_ROW_RELATION_SUBQUERY,
+      optimizeOneRowRelationSubqueryEnabled)
+    super.afterAll()
+  }
+
+  object Optimize extends RuleExecutor[LogicalPlan] {
+    val batches =
+      Batch("Subquery", Once,
+        OptimizeOneRowRelationSubquery,
+        PullupCorrelatedPredicates) :: Nil
+  }
+
+  private def assertHasDomainJoin(plan: LogicalPlan): Unit = {
+    assert(plan.collectWithSubqueries { case d: DomainJoin => d }.nonEmpty,
+      s"Plan does not contain DomainJoin:\n$plan")
+  }
+
+  val t0 = OneRowRelation()
+  val a = 'a.int
+  val b = 'b.int
+  val t1 = LocalRelation(a, b)
+  val t2 = LocalRelation('c.int, 'd.int)
+
+  test("Optimize scalar subquery with a single project") {
+    // SELECT (SELECT a) FROM t1
+    val query = t1.select(ScalarSubquery(t0.select('a)).as("sub"))
+    val optimized = Optimize.execute(query.analyze)
+    val correctAnswer = t1.select('a.as("sub"))
+    comparePlans(optimized, correctAnswer.analyze)
+  }
+
+  test("Optimize lateral subquery with a single project") {
+    Seq(Inner, LeftOuter, Cross).foreach { joinType =>
+      // SELECT * FROM t1 JOIN LATERAL (SELECT a, b)
+      val query = t1.lateralJoin(t0.select('a, 'b), joinType, None)
+      val optimized = Optimize.execute(query.analyze)
+      val correctAnswer = t1.select('a, 'b, 'a.as("a"), 'b.as("b"))
+      comparePlans(optimized, correctAnswer.analyze)
+    }
+  }
+
+  test("Optimize subquery with subquery alias") {
+    val inner = t0.select('a).as("t2")
+    val query = t1.select(ScalarSubquery(inner).as("sub"))
+    val optimized = Optimize.execute(query.analyze)
+    val correctAnswer = t1.select('a.as("sub"))
+    comparePlans(optimized, correctAnswer.analyze)
+  }
+
+  test("Optimize scalar subquery with multiple projects") {
+    // SELECT (SELECT a1 + b1 FROM (SELECT a AS a1, b AS b1)) FROM t1
+    val inner = t0.select('a.as("a1"), 'b.as("b1")).select(('a1 + 'b1).as("c"))
+    val query = t1.select(ScalarSubquery(inner).as("sub"))
+    val optimized = Optimize.execute(query.analyze)
+    val correctAnswer = Project(Alias(Alias(a + b, "c")(), "sub")() :: Nil, t1)
```

Comment:
nit: it's a bit weird that sometimes we use DSL and sometimes we use LogicalPlan directly. Can we be consistent? I think here can be `t1.select(('a + 'b).as("c").as("sub"))`

Reply:
The analyzer will remove the extra aliases which make the correct answer differs from the optimized. I will add a clean-up aliases rule in the test optimizer.